api/graphql/reservations/reservation_serializers.py

```diff
@@ -740,34 +744,154 @@ def check_metadata_fields(self, data, reservation_unit) -> None:
+    payment_type = ChoiceCharField(
+        choices=PaymentType.choices,
+        required=False,
+        help_text=(
+            "Type of the payment. "
+            f"Possible values are {', '.join(value[0].upper() for value in PaymentType.choices)}."
+        ),
+    )
+
     def __init__(self, *args, **kwargs):
         super().__init__(*args, **kwargs)
         # All fields should be read-only, except for the lookup
         # field (PK) which should be included in the input
         for field in self.fields:
             self.fields[field].read_only = True
         self.fields["pk"].read_only = False
+        self.fields["payment_type"].read_only = False
+
+    class Meta(ReservationUpdateSerializer.Meta):
+        fields = ["payment_type"] + ReservationUpdateSerializer.Meta.fields
+
+    def _requires_handling(self):
+        return (
+            self.instance.reservation_unit.filter(
+                require_reservation_handling=True
+            ).exists()
+            or self.instance.applying_for_free_of_charge
+        )
+
+    def _get_default_payment_type(self):
+        reservation_unit = self.instance.reservation_unit.first()
+        payment_types = reservation_unit.payment_types
+
+        if payment_types.count() == 0:
+            raise ValidationErrorWithCode(
+                "Reservation unit does not have payment types defined. At least one payment type must be defined.",
+                ValidationErrorCodes.INVALID_PAYMENT_TYPE,
+            )
+
+        # Rules to pick the default, defined in TILA-1974:
+        # 1. If only one payment type is defined, use that
+        # 2. If only INVOICE and ON_SITE are defined, use INVOICE
+        # 3. Otherwise use ONLINE
+
+        if payment_types.count() == 1:
+            return payment_types.first().code
+        elif (
+            payment_types.filter(
+                code__in=[PaymentType.INVOICE, PaymentType.ON_SITE]
+            ).count()
+            == payment_types.count()
+        ):
+            return PaymentType.INVOICE
+        else:
+            return PaymentType.ONLINE
+
+    def validate(self, data):
+        data = super().validate(data)
+
+        if self.instance.payment_order.exists():
+            raise ValidationErrorWithCode(
+                "Reservation cannot be changed anymore because it is attached to a payment order",
+                ValidationErrorCodes.CHANGES_NOT_ALLOWED,
+            )
+
+        if self.instance.reservation_unit.count() > 1:
+            raise ValidationErrorWithCode(
+                "Reservations with multiple reservation units are not supported.",
+                ValidationErrorCodes.MULTIPLE_RESERVATION_UNITS,
+            )
+
+        if not self._requires_handling():
+            payment_type = data.get("payment_type", "").upper()
+            reservation_unit = self.instance.reservation_unit.first()
+
+            if not reservation_unit.payment_product:
```

Comment:
This got me pondering; do we send every reservation_unit to webshop?

Reply:
No. Actually, we don't send any runits to webshop.

However, for each runit with **payment merchant** defined, we will create a product mapping - meaning runit has an UUID both we and webshop knows. When order is created, we will send this product UUID and merchant UUID as part of the order. These are the only information webshop knows about the runits.

This is why we check runit's product here - webshop payment is possible only for runits with product mapping defined. Hopefully this clarified things ☺️

Reply:
Yup, that clarified, thanks! Starting to fall off a bit of this webshop wagon